operations/security_and_authentication/Keycloak_User_Management_with_Kcadm.md

```diff
@@ -0,0 +1,305 @@
+# Keycloak User Management with kcadm.sh
+
+The Cray CLI requires a valid Keycloak JWT token.  If the token is not present it must be obtained by supplying valid user credentials at the time of CLI initialization.  During the installation or upgrade of a Shasta system the Keycloak user LDAP federation state or credentials may not be known or the external LDAP system for which the user may have been federated may not be available.  These scenarios make it impossible to initialize the Cray CLI with a previously valid Keycloak user or LDAP federated Keycloak user.  If for whatever reason the Keycloak UI is not available to the admin to verify the state of the user, using Keycloak's kcadm.sh utility from the command line will be required.
+
+The kcadm.sh utility is included with the Shasta Keycloak container image and can be used as described below to perform troubleshooting operations that can verify the existence of a user, determine if the user is federated, change the user credential (for a keycloak-local account only) or create a new local Keycloak user that is compatible (has the correct role mappings) with the Cray CLI.  All the operations described below must be run as root from the ncn-m001 node.
+
+Usage of the bash function(s) provided below will require addition into the Linux environment.  This can be accomplished by pasting the function(s) into your shell or pasting the function(s) into a file and then sourcing the file before running the indicated function(s).
+
+## Checking the Status of a Keycloak User
+
+Use this procedure to determine if a user both exists in Keycloak and if the user is LDAP federated or not.  If the user is LDAP federated and the credential is not known, update the user credential in LDAP.  It will not be possible to use a federated user account to initialize the Cray CLI if LDAP is not available.  In this case you may create a local Keycloak user for use with the Cray CLI by following the procedure below for creating a local Keycloak Cray CLI user.  The credential for a local account may be updated if needed by following the procedure below for updating a local Keycloak account credential.
```

Comment:
```suggestion
Use this procedure to determine if a user both exists in Keycloak and if the user is LDAP federated or not.  If the user is LDAP federated and the credential is not known, update the user credential in LDAP.  It will not be possible to use a federated user account to initialize the Cray CLI if LDAP is not available.  In this case, create a local Keycloak user for use with the Cray CLI by following the procedure below for creating a local Keycloak Cray CLI user.  The credential for a local account may be updated if needed by following the procedure below for updating a local Keycloak account credential.
```